Config.md
# Configuration options

Echidna's CLI can be used to choose the contract to test and load a
configuration file.

```
$ echidna-test contract.sol --contract TEST --config config.yaml
```

The configuration file allows users to choose EVM and test generation
parameters. An example of a complete and annotated config file with the default
options can be found at
[examples/solidity/basic/default.yaml](examples/solidity/basic/default.yaml).

## Options in the configuration file

Echidna allows a variety of config options in a yaml file (e.g `config.yaml`):

* `testLimit`
  * Type: Int
  * Default: `50000`
  * Description: Number of sequences of transactions to generate during testing.  

* `seqLen` 
  * Type: Int
  * Default: `100`
  * Description: Number of transactions to generate during testing.

* `shrinkLimit` 
  * Type: Int
  * Default: `5000`
  * Description: Number of tries to attempt to shrink a failing sequence of transactions.

* `contractAddr`
  * Type:  Address
  * Default: `"0x00a329c0648769a73afac7f9381e08fb43dbea72"`
  * Description: Address to deploy the contract to test.

* `coverage`
  * Type:  Bool
  * Default: `true`
  * Description: Enable the use of coverage-guided fuzzing and corpus collection.

* `corpusDir`
  * Type:  String
  * Default: `null`
  * Description: Directory to save the corpus collected (requires coverage enabled).

* `deployer`
  * Type:  Address
  * Default: `"0x00a329c0648769a73afac7f9381e08fb43dbea70"`
  * Description: Address of the deployer of the contract to test.

* `sender`
  * Type: [Address]
  * Default: `["0x00a329c0648769a73afac7f9381e08fb43dbea70"]`
  * Description: List of addresses to (randomly) use during for the transactions sent during testing.

* `psender`
  * Type: Address
  * Default: `"0x00a329c0648769a73afac7f9381e08fb43dbea70"`
  * Description: Address of the sender of the property to test.

* `prefix`
  * Type: String 
  * Default: `"echidna_"`
  * Description: Prefix of the function names used as properties in the contract to test. 

* `propMaxGas`
  * Type: Int 
  * Default: `12500000` (current max gas per block)
  * Description: Maximum amount of gas to consume when running function properties.

* `testMaxGas`
  * Type: Int 
  * Default: `12500000` (current max gas per block)
  * Description: Maximum amount of gas to consume when running random transactions.

* `maxGasprice`
  * Type: Int 
  * Default: `0` 
  * Description: Maximum amount of gas price to randomly use in transactions. Do not change it unless you absolutely need it.

* `maxTimeDelay`
  * Type: Int 
  * Default: `604800`  (one week) 
  * Description: Maximum amount of seconds of delay between transactions.

* `maxBlockDelay`
  * Type: Int 
  * Default: `60480`   
  * Description: Maximum amount of block numbers between transactions.
 
* `solcArgs`
  * Type: [String] 
  * Default: `[]`
  * Description: Additional arguments to use in `solc` for the compilation of the contract to test. 

* `cryticArgs`
  * Type: [String] 
  * Default: `[]`
  * Description: Additional arguments to use in `crytic-compile` for the compilation of the contract to test. 

* `quiet`
  * Type: Bool
  * Default: `False`
  * Description: Hide `solc` stderr output and additional information during the testing.

* `dashboard`
  * Type: Bool
  * Default: `True`
  * Description: Show the ncurses dashboard with real-time information on the properties to test during the fuzzing campaign.

* `initialBalance`
  * Type: Int
  * Default: `0xffffffff`
  * Description: Initial Ether balance of `deployer` and each of the `sender` accounts.

* `format`
  * Type: String
  * Default: `"text"`
  * Description: Select an UI to show the results of each test. Either`"text", `"json"` or`"none"`.

Echidna supports three different output drivers. There is the default `text`
driver, a `json` driver, and a `none` driver, which should suppress all
`stdout` output. 

## JSON output

The JSON driver reports the overall campaign as follows.

```json
Campaign = {
  "success"      : bool,
  "error"        : string?,
  "tests"        : [Test],
  "seed"         : number,
  "coverage"     : Coverage,
  "gas_info"     : [GasInfo]
}
Test = {
  "contract"     : string,
  "name"         : string,
  "status"       : string,
  "error"        : string?,
  "testType"     : string,
  "transactions" : [Transaction]?
}
Transaction = {
  "contract"     : string,
  "function"     : string,
  "arguments"    : [string]?,
  "gas"          : number,
  "gasprice"     : number
}
```

`Coverage` is a dict describing certain coverage increasing calls.
Each `GasInfo` entry is a tuple that describes how maximal
gas usage was achieved, and also not too important. These interfaces are
subject to change to be slightly more user friendly at a later date. `testType`
will either be `property` or `assertion`, and `status` always takes on either
`fuzzing`, `shrinking`, `solved`, `passed`, or `error`.


CHANGELOG.md
## TODO

## 1.7.1

* Tweaked mutators and improved test stability (#628)
* Documented known issues and limitations (#655)
* Automatically initialize addresses when used (#657)
* Improved coverage to count number of EVM frames (#624)
* Avoid mutations to generate inputs outside their expected ABI range (#650)
* Various small to run Vyper contracts (#645)
* Improved integer generation to use small values (#644)
* Added two CLI options: --corpus-dir and --check-asserts (#640)
* Refactored shrinkSeq to improve readability (#639)
* Fixed UI to fit long function calls (#635)
* Refactor coverage types and added corpus size in UI (#627)
* Fixed link to macOS binary in binaries.soliditylang.org (#629)
* Fixed default.nix to use 1.7.0 as version (#623)
* Refactored Test type (#622) 

## 1.7.0

* Refactored and improved etheno support to be more useful (#615) 
* Coverage filenames are not overwritten (#620)
* Refactored the mutator code (#618)
* More corpus and array mutations implemented (#372)
* Source coverage is printed after fuzzing campaign (#516)
* Nix improvements and fixes (#603, #604, #608, #612)
* Simplified slither information parsing (#543) 
* Enabled use of coverage by default (#605)
* Run echidna tests in parallel (#571)

## 1.6.1

* Use a sensible default value for `block.gaslimit` (#596)
* Use metadata to detect deployed contracts (#593)
* Semver integration for improving testing with different solc versions (#594)
* Added some performance improvement in property execution (#576)
* Fixed wait bug when shrinking (#584) 
* Added funwithnumber example from Sabre (#565)
* Improved function filtering to be more precise (#570)
* Small fixes in the macOS CI (#597), the README (#590) and Nix (#581) 

## 1.6.0
* Slither is now a required dependency.
* Made sure that <ESC> gets mapped correctly, fixing #474 (#503)
* Fix library timestamp/block delay issue (#510)
* Make stack limit exceeded a revert (#517)
* Add some default transaction constants (#532)
* Make large constants work better with dictionaries (#523)
* Fix "flanky" corpus tests (#537)
* HEVM updated to [150dddc67b6cbad75fd4ae5a689452892f55ea26] (0.42) (#511)
  * this may introduce a slight performance hit as everything is partially symbolic
* Improved shrinking and pretty printing (#518)
* Integrate slither results (#451)
* Disagreements about `MonadState` and `MonadReader` (#545, #513)
* Simplified a bunch of code (#548, #549, #513)
* Fix negative address bug (#552)
* Various Github Actions improvements (#527, #554)
* Allow to bypass EIP-170 and set up a custom max code size (#544)
 
## 1.5.1

* Fix timestamp and block delays having the initial timestamp/block added to them (#460, #469)
* Fix GitHub actions due to new macOS image (#468)
* Relax `package.yaml` constraints (#466)
* Fix `extcodesize` by creating contracts correctly (#453, #454, #465)
* Refactor unit tests (#457)
* Add seed to UI (#456)
* Set an initial timestamp and block number (#455)
* Better README (#450)
* Code simplifications (#448)
* Detect contract deployment failures (#447)
* DockerHub containers in addition to GitHub (#445)
* Brought back print reporting after exiting the UI (#441)
* Refactor `Tx` (#437)

## 1.5.0

* Switch to x.y.z versioning scheme
* Refactor campaign-related code (#384)
* Fixed hevm-related memory leak (#423)
* Fixed array mutation memory leak (#442)
* Benchmark mode to run without tests (#420, #409)
* Migrated to GitHub actions (#414, #402)
* Corpus mutation handling fixes (#395, #403)
* Improved JSON output and reworked UI handling to better handle interactive and non-interactive mode (#405)
* Use safe mutation parameters to favor purely random generation (#393)
* Show address in hex using "0x" (#386)
* Enable addresses shrinking (#385)
* Fixed coverage counting when a contract is deployed multiple times (#410)
* Prioritization to mutate each list of transactions from the corpus (#376, #395)
* Improved array generation using generic mutations (#357)
* Multiple coverage tracking according to the result of a transaction (#369)
* Allow to control corpus mutation constans using a configuration file (#373)

## 1.4.0.1
* HEVM updated to [af84e2ee0a0654fdaa91186384233cf1731ee7ce]

## 1.4.0.0

* Filtering functions either blacklisting or whitelisting functions to call during a fuzzing campaign (#341)
* Support to collect, save and load a corpus of transactions in JSON (#352, #353)
* Basic mutations based on collected corpus (#370)
* Use of `hlint` as Github action to verify new PR (#366)

## 1.3.0.0

* Gaining knowledge of contract-created contract addresses (#295)
* Improved `crytic-compile` error handling when missing (#335)
* Support testing from arbitrary blockchain state via etheno (#333)
* Some shrinking improvements (#336)
* UI improvements (#339)
* UI updates are now threaded separately from fuzzing campaign (#345)
* `Tx` rework to unify `Echidna.RPC` and `Transaction` (#346)
* Adjustment of testsuite parameters (#347)
* Multi-ABI support by fuzzing all known ABIs (#344)
* Fix indefinite `MVar` blocking (#350)
* AddressArrayExample (#348)
* `Tx` JSON conversions (#352)
* Smaller docker container, multi-stage build (#349)
* Gas worst case estimation (#355)

## 1.2.0.0

* Warn on unused config keys (#301)
* `default.yaml` kept up to date via CI checks (#301)
* Improved shrinking (#310)
* Solidity ABIv2 support (#284, #303, #317)
* Early termination on property falure (#323)
* Timeouts (#320)
* HEVM updated to [2cc059b49cae613025b925f0273b906e25484b68]
* Dictionary control (#312, #321)
* Gas price fuzzing (#307)
* Coverage bug fixed (#300)
* Contract creation bug fixed (#293)
* New startup message (#290)
* Add `--version` flag (#285)
* Support time/block delays (#282)
* Support `crytic` arguments (#280)
* Detect assertion failure (#276)
* Use return values as constants (#262)
* Reproducible testing with seeds (#254)

## 1.0.0.0

* Initial stable release


lib/Echidna/Config.hs
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TupleSections #-}

module Echidna.Config where

import Control.Lens
import Control.Monad.Catch (MonadThrow)
import Control.Monad.IO.Class (MonadIO(..))
import Control.Monad.Reader (Reader, ReaderT(..), runReader)
import Control.Monad.State (StateT(..), runStateT)
import Control.Monad.Trans (lift)
import Data.Bool (bool)
import Data.Aeson
import Data.Has (Has(..))
import Data.HashMap.Strict (keys)
import Data.HashSet (HashSet, fromList, insert, difference)
import Data.Maybe (fromMaybe)
import Data.Text (Text, isPrefixOf)
import EVM (result)
import EVM.Concrete (Word(..), Whiff(..))

import qualified Control.Monad.Fail as M (MonadFail(..))
import qualified Data.ByteString as BS
import qualified Data.List.NonEmpty as NE
import qualified Data.Yaml as Y

import Echidna.Solidity
import Echidna.Test
import Echidna.Types.Campaign (CampaignConf(CampaignConf))
import Echidna.Mutator.Corpus (defaultMutationConsts)
import Echidna.Types.Tx  (TxConf(TxConf), maxGasPerBlock, defaultTimeDelay, defaultBlockDelay)
import Echidna.UI
import Echidna.UI.Report

-- | Our big glorious global config type, just a product of each local config.,
data EConfig = EConfig { _cConf :: CampaignConf
                       , _nConf :: Names
                       , _sConf :: SolConf
                       , _tConf :: TestConf
                       , _xConf :: TxConf
                       , _uConf :: UIConf
                       }
makeLenses ''EConfig

data EConfigWithUsage = EConfigWithUsage { _econfig   :: EConfig
                                         , _badkeys   :: HashSet Text
                                         , _unsetkeys :: HashSet Text
                                         }
makeLenses ''EConfigWithUsage

instance Has EConfig EConfigWithUsage where
  hasLens = econfig

instance Has CampaignConf EConfig where
  hasLens = cConf

instance Has Names EConfig where
  hasLens = nConf

instance Has SolConf EConfig where
  hasLens = sConf

instance Has TestConf EConfig where
  hasLens = tConf

instance Has TxConf EConfig where
  hasLens = xConf

instance Has UIConf EConfig where
  hasLens = uConf

instance FromJSON EConfig where
  -- retrieve the config from the key usage annotated parse
  parseJSON = fmap _econfig . parseJSON

instance FromJSON EConfigWithUsage where
  -- this runs the parser in a StateT monad which keeps track of the keys
  -- utilized by the config parser
  -- we can then compare the set difference between the keys found in the config
  -- file and the keys used by the parser to comopute which keys were set in the
  -- config and not used and which keys were unset in the config and defaulted
  parseJSON o = do
    let v' = case o of
                  Object v -> v
                  _        -> mempty
    (c, ks) <- runStateT (parser v') $ fromList []
    let found = fromList (keys v')
    return $ EConfigWithUsage c (found `difference` ks) (ks `difference` found)
    -- this parser runs in StateT and comes equipped with the following
    -- equivalent unary operators:
    -- x .:? k (Parser) <==> x ..:? k (StateT)
    -- x .!= v (Parser) <==> x ..!= v (StateT)
    -- tl;dr use an extra initial . to lift into the StateT parser
    where parser v =
            let useKey k = hasLens %= insert k
                x ..:? k = useKey k >> lift (x .:? k)
                x ..!= y = fromMaybe y <$> x
                getWord s d = C Dull . fromIntegral <$> v ..:? s ..!= (d :: Integer)

                -- TxConf
                xc = TxConf <$> getWord "propMaxGas" maxGasPerBlock
                            <*> getWord "testMaxGas" maxGasPerBlock
                            <*> getWord "maxGasprice" 0
                            <*> getWord "maxTimeDelay" defaultTimeDelay     
                            <*> getWord "maxBlockDelay" defaultBlockDelay
                            <*> getWord "maxValue" 100000000000000000000 -- 100 eth

                -- TestConf
                tc = do
                  psender <- v ..:? "psender" ..!= 0x00a329c0648769a73afac7f9381e08fb43dbea70
                  fprefix <- v ..:? "prefix"  ..!= "echidna_"
                  let goal fname = if (fprefix <> "revert_") `isPrefixOf` fname then ResRevert else ResTrue
                      classify fname vm = maybe ResOther classifyRes (vm ^. result) == goal fname
                  return $ TestConf classify (const psender)

                -- CampaignConf
                cov = v ..:? "coverage" <&> \case Just False -> Nothing
                                                  _          -> Just mempty
                cc = CampaignConf <$> v ..:? "testLimit"   ..!= 50000
                                  <*> v ..:? "stopOnFail"  ..!= False
                                  <*> v ..:? "estimateGas" ..!= False
                                  <*> v ..:? "seqLen"      ..!= 100
                                  <*> v ..:? "shrinkLimit" ..!= 5000
                                  <*> cov
                                  <*> v ..:? "seed"
                                  <*> v ..:? "dictFreq"    ..!= 0.40
                                  <*> v ..:? "corpusDir"   ..!= Nothing
                                  <*> v ..:? "mutConsts"   ..!= defaultMutationConsts

                -- SolConf
                defaultAddr     = 0x00a329c0648769a73afac7f9381e08fb43dbea72
                defaultDeployer = 0x00a329c0648769a73afac7f9381e08fb43dbea70
                fnFilter = bool Whitelist Blacklist <$> v ..:? "filterBlacklist" ..!= True
                                                    <*> v ..:? "filterFunctions" ..!= []
                sc = SolConf <$> v ..:? "contractAddr"    ..!= defaultAddr
                             <*> v ..:? "deployer"        ..!= defaultDeployer
                             <*> v ..:? "sender"          ..!= (0x10000 NE.:| [0x20000, defaultDeployer])
                             <*> v ..:? "balanceAddr"     ..!= 0xffffffff
                             <*> v ..:? "balanceContract" ..!= 0
                             <*> v ..:? "codeSize"        ..!= 0x6000      -- 24576 (EIP-170)
                             <*> v ..:? "prefix"          ..!= "echidna_"
                             <*> v ..:? "cryticArgs"      ..!= []
                             <*> v ..:? "solcArgs"        ..!= ""
                             <*> v ..:? "solcLibs"        ..!= []
                             <*> v ..:? "quiet"           ..!= False
                             <*> v ..:? "initialize"      ..!= Nothing
                             <*> v ..:? "multi-abi"       ..!= False
                             <*> v ..:? "checkAsserts"    ..!= False
                             <*> v ..:? "benchmarkMode"   ..!= False
                             <*> fnFilter
                names :: Names
                names Sender = (" from: " ++) . show
                names _      = const ""
                mode = fromMaybe Interactive <$> (v ..:? "format" >>= \case
                  Just ("text" :: String) -> pure . Just . NonInteractive $ Text
                  Just "json"             -> pure . Just . NonInteractive $ JSON
                  Just "none"             -> pure . Just . NonInteractive $ None
                  Nothing -> pure Nothing
                  _ -> M.fail "unrecognized format type (should be text, json, or none)") in
            EConfig <$> cc <*> pure names <*> sc <*> tc <*> xc
                    <*> (UIConf <$> v ..:? "timeout" <*> mode)

-- | The default config used by Echidna (see the 'FromJSON' instance for values used).
defaultConfig :: EConfig
defaultConfig = either (error "Config parser got messed up :(") id $ Y.decodeEither' ""

-- | Try to parse an Echidna config file, throw an error if we can't.
parseConfig :: (MonadThrow m, MonadIO m) => FilePath -> m EConfigWithUsage
parseConfig f = liftIO (BS.readFile f) >>= Y.decodeThrow

-- | Run some action with the default configuration, useful in the REPL.
withDefaultConfig :: ReaderT EConfig m a -> m a
withDefaultConfig = (`runReaderT` defaultConfig)

-- | 'withDefaultConfig' but not for transformers
withDefaultConfig' :: Reader EConfig a -> a
withDefaultConfig' = (`runReader` defaultConfig)
